models/contacts.js

```diff
@@ -1,19 +1,126 @@
-// const fs = require('fs/promises')
+const { readFile, writeFile } = require("fs/promises");
+const { v4: uuidv4 } = require("uuid");
+const path = require("path");
+require("colors");
 
-const listContacts = async () => {}
+const baseDir = __dirname;
+const fileName = "contacts.json";
+const contactsPath = path.join(baseDir, fileName);
 
-const getContactById = async (contactId) => {}
+const contactsArray = () => {
+  return readFile(contactsPath, "utf-8")
+    .then((data) => JSON.parse(data))
+    .catch((error) => console.log(error.message.red));
+};
 
-const removeContact = async (contactId) => {}
+const saveContactArray = (contacts) => {
+  writeFile(contactsPath, JSON.stringify(contacts, null, 2), (error) => {
+    if (error) {
+      console.error(error.message.red);
+    }
+  });
+};
 
-const addContact = async (body) => {}
+const listContacts = async () => {
+  try {
+    const contacts = await contactsArray();
+    return contacts;
+  } catch (error) {
+    console.log(error.message.red);
+    throw error;
+  }
+};
 
-const updateContact = async (contactId, body) => {}
+const getContactById = async (contactId) => {
+  try {
+    const contacts = await contactsArray();
+    const index = contacts.findIndex((contact) => contact.id === contactId);
+    if (index === -1) {
+      throw new Error(`The contact with id number ${contactId} does not exist`);
+    } else {
+      return contacts[index];
+    }
+  } catch (error) {
+    console.log(error.message.red);
+    throw error;
+  }
+};
+
+const removeContact = async (contactId) => {
+  try {
+    const contacts = await contactsArray();
+    const index = contacts.findIndex((contact) => contact.id === contactId);
+    if (index === -1) {
+      throw new Error(`The contact with id number ${contactId} does not exist`);
+    } else {
+      const deleteContact = contacts.splice(index, 1);
+      await saveContactArray(contacts);
+      return deleteContact[0];
+    }
+  } catch (error) {
+    console.log(error.message.red);
+    throw error;
+  }
+};
+
+const addContact = async (body) => {
+  try {
+    const { name, email, phone } = body;
+    const contacts = await contactsArray();
+    const findContact = contacts.findIndex(
+      (contact) =>
+        contact.name === name ||
+        contact.email === email ||
+        contact.phone === phone
+    );
+    if (findContact !== -1) {
+      throw new Error("Contact with this data already exists");
```

Comment:
Myślę że mogą być osoby które mają to samo imię i nazwisko, ale już numer powinien być unique. To tak na marginesie.

Reply:
a no racja, zmieniłem na `contact.email === email || contact.phone === phone`